src/Avalonia.Base/Styling/Activators/ContainerQueryActivatorBase.cs

```diff
@@ -0,0 +1,83 @@
+﻿using System;
+using System.Linq;
+using Avalonia.Layout;
+using Avalonia.VisualTree;
+
+namespace Avalonia.Styling.Activators
+{
+    internal abstract class ContainerQueryActivatorBase : StyleActivatorBase, IStyleActivatorSink
+    {
+        private readonly Visual _visual;
+        private readonly string? _containerName;
+        private Layoutable? _currentScreenSizeProvider;
+
+        public ContainerQueryActivatorBase(
+            Visual visual, string? containerName = null)
+        {
+            _visual = visual;
+            _containerName = containerName;
+
+            _visual.AttachedToVisualTree += Visual_AttachedToVisualTree;
+        }
+
+        private void Visual_AttachedToVisualTree(object? sender, VisualTreeAttachmentEventArgs e)
+        {
+            InitializeScreenSizeProvider();
+        }
+
+        protected Layoutable? CurrentContainer => _currentScreenSizeProvider;
+
+        void IStyleActivatorSink.OnNext(bool value) => ReevaluateIsActive();
+
+        protected override void Initialize()
+        {
+            InitializeScreenSizeProvider();
+        }
+
+        protected override void Deinitialize()
+        {
+            _visual.AttachedToVisualTree -= Visual_AttachedToVisualTree;
+
+            if (_currentScreenSizeProvider is { } && Container.GetQueryProvider(_currentScreenSizeProvider) is { } provider)
+            {
+                provider.WidthChanged -= WidthChanged;
+                provider.HeightChanged -= HeightChanged;
+                _currentScreenSizeProvider = null;
+            }
+        }
+
+        private void InitializeScreenSizeProvider()
+        {
```

Comment:
This method is called every time the visual is attached to the tree. It means switching between visual trees will add new handlers to the `WidthChanged`/`HeightChanged` events each time without ever removing the previous one.

`_visual.AttachedToVisualTree` should probably be unsubscribed from in this method, until it's detached. Then, on detach, remove the `Width/Height` handlers.

`Deinitialize()` should remove everything subscribed to so far unconditionally.

To reproduce in ControlCatalog, select the Container Queries tab, then another one, then the Container Queries tab again.